src/commands/mod.rs

```diff
@@ -23,3 +30,55 @@ impl Command {
         }
     }
 }
+
+/// A command that can be invoked in an LDNS compatibility mode
+///
+/// These commands do their own argument parsing, because clap cannot always
+/// (easily) parse arguments in the same way that the ldns tools do.
+///
+/// The `LdnsCommand::parse_ldns` function should parse arguments obtained
+/// with [`std::env::args`] or [`std::env::args_os`] and return an error in
+/// case of invalid arguments. The help string provided as
+/// [`LdnsCommand::HELP`] is automatically appended to returned errors.
+pub trait LdnsCommand: Into<Command> {
+    const HELP: &'static str;
+
+    fn parse_ldns() -> Result<Self, Error>;
+
+    fn parse_ldns_args() -> Result<Args, Error> {
+        match Self::parse_ldns() {
+            Ok(c) => Ok(Args::from(c.into())),
+            Err(e) => Err(format!("Error: {e}\n\n{}", Self::HELP).into()),
+        }
+    }
+
+    /// Utility function to parse an [`OsStr`] with a custom function
+    fn parse_os_with<T, E>(
+        opt: &str,
+        val: &OsStr,
+        f: impl Fn(&str) -> Result<T, E>,
+    ) -> Result<T, Error>
+    where
+        E: std::fmt::Display,
+    {
+        let Some(s) = val.to_str() else {
+            return Err(format!("Invalid value for {opt}: {val:?} is not valid unicode",).into());
```

Comment:
I wonder could this be improved to know what the last arg parsed by `parser.next()` was? To save boiler plate repeating the argument letter?

Reply:
This could be improved later, it shouldn't block this PR.